### DISCUSSION TITLE
wire:loading for wire:navigate

### DISCUSSION THREAD

**Discussion by angus-mcritchie** [livewire/livewire#7578] [3. Feature ideas]
While using `wire:navigate`, it would be great to show a loading state while the next page is loading.

Perhaps add a new directive called `wire:navigating`, which acts just like `wire:loading` but for redirects.

I did look for something in the docs and other discussions for this but couldn't find anything but this [old Reddit thread](https://www.reddit.com/r/laravel/comments/t4kk4j/livewire_loading_while_redirecting/).

```blade
<a href="/slow-page" wire:navigate>
    <span wire:navigating> {{-- new wire:navigating directive --}}
        Loading page...
    </span>

    <span wire:navigating.remove> {{-- new wire:navigating directive --}}
        Load Slow Page
    </span>
</a>
```

The implementation above is just an idea; other implementations would be okay, such as using `wire:loading` with `wire:target` as the URL.

Also, it would be ideal if this feature works with the Livewire component $this->redirect() (assuming it's using `wire:navigate`).

This would also be a nice way to solve this [wire:loading issue](https://github.com/livewire/livewire/discussions/7369) as you could show "Logging in...", then "Redirecting..." messages.

Please let me know if there's a way to do this already 🙏

First post here; any feedback you can give is welcome.

**Comment by joshhanley** [comment]
@angus-mcritchie makes sense to me. PRs are welcome 🙂 ideally it would be better to use `wire:loading` some how. Currently there is a loading bar that is displayed at the top of the page. Maybe you can look at how that is implemented to see how this feature might work.

**Comment by kaburagisec** [comment]
hello, I implemented this with wire:loading and wire:click, where you can direct wire:click to a function that doesn't execute anything. example:

[dumper-index.blade.php]
```
<a wire:click="nothing('{{$dumper->id}}')" href="{{route('dumper.detail', ['name' => $dumper->name])}}" wire:navigate><button wire:loading.class="opacity-50 disabled" wire:target="nothing('{{$dumper->id}}')" type="button" class="btn btn-primary btn-sm w-100"><i class="fa-solid fa-eye"></i> Info</button></a><br>
```

[DumperIndex.php]
```
public function nothing($id)
{
        //nothing happen here
}
```

with the code above, the button will receive the loading attribute and still direct the user to the desired url/route with wire:navigate. I hope this helps!

**Comment by kaburagisec** [comment]
```
<a href="/slow-page" wire:navigate>
    <span wire:navigating> {{-- new wire:navigating directive --}}
        Loading page...
    </span>

    <span wire:navigating.remove> {{-- new wire:navigating directive --}}
        Load Slow Page
    </span>
</a>
```

This code can be implemented by:

```
<a wire:click="nothing('unique_id')" href="/slow-page" wire:navigate>
    <span wire:target="nothing('unique_id')" wire:loading> {{-- new wire:navigating directive --}}
        Loading page...
    </span>

    <span wire:target="nothing('unique_id')" wire:loading.remove> {{-- new wire:navigating directive --}}
        Load Slow Page
    </span>
</a>
```

I hope this helps!

**Comment by adriansio** [comment]
I encountered this issue too and the UI feels unpolished because Livewire halts the loading state when it returns the redirect but at the same time the element that had `wire:loading.remove` doesn't reappear.

I feel this is quite a common scenario, are there any chances that this can be fixed at the framework level?

**Comment by levu42** [comment]
What about the syntax `<span wire:loading wire:target="wire:navigate">...</span>`? That looks to me like the cleanest implementation, it adds no new syntax apart from a special word as target.

**Reply by angus-mcritchie** [reply]
I do like this syntax @levu42 but I wonder if we should allow targeting specific URLs?

If you've got a list of links and want to display a loading indicator for each, they will all render loading when any of the links are clicked 🤔

Or even what element triggered the navigation (if any, as it can be triggered from the Livewire JS directly).

**Reply by patriktoth67** [reply]
I went down that rabbit hole already, made a PR but ended up leaving it as not worth to do. Here is the first PR: #8118 which needed to be changed after v3.4.9: #8165

At the end Caleb decided to close it, and offered a solution:
#https://github.com/livewire/livewire/pull/8165#issuecomment-2025072603

It's easier to do it with that instead of a new feature.

**Comment by adriansio** [comment]
That could be good. Livewire is amazing and V3 has moved really forward in terms of speed and UI snappy feel, but this is some very annoying "glitch" that feels unpolished.

When using full page components, redirects are quite common so it will be great to have this update.

@joshhanley @calebporzio – apologize for pinging, just wondering if the core team is aware of this. Thanks so much!

**Reply by joshhanley** [reply]
@adriansio yep I'm aware, see [my comment](https://github.com/livewire/livewire/discussions/7578#discussioncomment-7903386) at the top of this thread. PRs are always welcome 🙂

**Comment by masihgh** [comment]
hello and thanks for everything!
i want thius feature at every project i have
somthing like wire:loading-navigate
please please do that!

**Comment by abbasgholia** [comment]
I have the same problem, please someone help us.

**Comment by angus-mcritchie** [comment]
Firstly, thanks to @patriktoth67 for all his work on this in his [PR](https://github.com/livewire/livewire/pull/8165).

For those who hit this page and want a solution, @calebporzio [suggested a solution](https://github.com/livewire/livewire/pull/8165#issuecomment-2025072603).

## Solution
Here is my simple implementation using AlpineJS and Livewire navigate events.

### Setup
Add a [Alpine JS store](https://alpinejs.dev/globals/alpine-store) called navigate:
```javascript
Alpine.store('navigate', {
    loading: false,
    to: null
});

document.addEventListener('livewire:navigate', event => {
    Alpine.store('navigate').loading = true;
    Alpine.store('navigate').to = event.detail.url.href;
});

document.addEventListener('livewire:navigated', event => {
    Alpine.store('navigate').loading = false;
    Alpine.store('navigate').to = null;
});
```
If you're not sure how to do this with your Livewire installation, see the [Livewire docs on loading Alpine JS](https://livewire.laravel.com/docs/alpine#manually-bundling-alpine-in-your-javascript-build).
### Show loading indicator on any navigation
```blade
<a href="{{ route('common-page') }}" wire:navigate>
    Common Page
    <span x-data x-show="$store.navigate.to !== null">Loading...</span>
</a>
```

### Show loading indicator for a specific URL
```blade
<a href="{{ route('slow-page') }}" wire:navigate>
    Slow Page
    <span x-data x-show="$store.navigate.to === '{{ route('slow-page') }}'">Loading...</span>
</a>
```

Thanks again for everyone's help on this discussion ❤️

**Reply by abbasgholia** [reply]
excellent
it worked for me
thanks

**Reply by jeffegiovani** [reply]
Thanks, it works!

**Reply by hasan-aftab** [reply]
Thanks mate

      x-init="
         window.addEventListener('livewire:navigate', () => isNavigating = true);
         window.addEventListener('livewire:navigated', () => isNavigating = false);
         "

**Comment by hasan-aftab** [comment]
For full page livewire component, u may can use livewire v3 features to do that:

`    Route::get('profile', Profile::class)->lazy()->name('settings.profile');
`

In profile component:

```
   public function placeholder()
    {
        return view('components.profile-shimmer');
    }
```
